manager/controllers/scaling/metrics_cache.go

```diff
@@ -0,0 +1,217 @@
+package scaling
+
+import (
+	"container/list"
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	dfv1 "github.com/argoproj-labs/argo-dataflow/api/v1alpha1"
+	lru "github.com/hashicorp/golang-lru"
+	pmodel "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	httpClient   *http.Client
+	metricsCache *lru.Cache
+)
+
+func init() {
+	t := http.DefaultTransport.(*http.Transport).Clone()
+	t.MaxIdleConns = 32
+	t.MaxConnsPerHost = 32
+	t.MaxIdleConnsPerHost = 32
+	t.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+	httpClient = &http.Client{Timeout: 1 * time.Second, Transport: t}
+
+	metricsCache, _ = lru.New(1000)
+}
+
+type MetricsCacheHandler struct {
+	client client.Client
+
+	stepMap  map[string]*list.Element
+	stepList *list.List
+	lock     sync.RWMutex
+	workers  int
+}
+
+func NewMetricsCacheHandler(client client.Client, workers int) *MetricsCacheHandler {
+	return &MetricsCacheHandler{
+		client:  client,
+		workers: workers,
+
+		stepMap:  make(map[string]*list.Element),
+		stepList: list.New(),
+		lock:     sync.RWMutex{},
+	}
+}
+
+func (m *MetricsCacheHandler) Contains(step *dfv1.Step) (bool, error) {
+	key, err := cache.MetaNamespaceKeyFunc(step)
+	if err != nil {
+		return false, fmt.Errorf("failed to get key for step object: %w", err)
+	}
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	_, ok := m.stepMap[key]
+	return ok, nil
+}
+
+func (m *MetricsCacheHandler) Length() int {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	return m.stepList.Len()
+}
+
+func (m *MetricsCacheHandler) StartWatchingStep(step *dfv1.Step) error {
+	key, err := cache.MetaNamespaceKeyFunc(step)
+	if err != nil {
+		return err
+	}
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if _, ok := m.stepMap[key]; !ok {
+		m.stepMap[key] = m.stepList.PushBack(step)
+	}
+	return nil
+}
+
+func (m *MetricsCacheHandler) StopWatchingStep(step *dfv1.Step) error {
+	key, err := cache.MetaNamespaceKeyFunc(step)
+	if err != nil {
+		return err
+	}
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if e, ok := m.stepMap[key]; ok {
+		_ = m.stepList.Remove(e)
+		delete(m.stepMap, key)
+	}
+	return nil
+}
+
+func (m *MetricsCacheHandler) Start(ctx context.Context) {
+	stepCh := make(chan *dfv1.Step)
```

Comment:
can we use a `workqueue` instead, it'll de-dupe steps - you should queue the `key`, not the step when you do that (as steps might not be unqiu)

Reply:
It's already de-duped by the key, guaranteed by the map. I don't need to queue the `step`, doing that is simply because underlining functions accepts step as parameters.